nominal/timedomain.py

```diff
@@ -0,0 +1,151 @@
+""" """
+
+from __future__ import annotations
+
+import warnings
+from dataclasses import dataclass
+from datetime import datetime
+from types import MappingProxyType
+from typing import Literal, Mapping
+
+import numpy as np
+from typing_extensions import TypeAlias
+
+from ._api.combined import ingest_api
+from ._timeutils import IntegralNanosecondsUTC
+
+
+@dataclass(frozen=True)
+class Iso8601:
+    pass
+
+
+@dataclass(frozen=True)
+class Epoch:
+    unit: _LiteralTimeUnit
+
+
+@dataclass(frozen=True)
+class Relative:
+    unit: _LiteralTimeUnit
+    start: datetime | IntegralNanosecondsUTC
+    """The starting time to which all relatives times are relative to."""
+
+
+@dataclass(frozen=True)
+class Custom:
+    format: str
+    default_year: int | None = None
+
+
+ISO_8601 = Iso8601()
+EPOCH_NANOSECONDS = Epoch("nanoseconds")
+EPOCH_MICROSECONDS = Epoch("microseconds")
+EPOCH_MILLISECONDS = Epoch("milliseconds")
+EPOCH_SECONDS = Epoch("seconds")
+EPOCH_MINUTES = Epoch("minutes")
+EPOCH_HOURS = Epoch("hours")
+
+_LiteralTimeUnit: TypeAlias = Literal[
+    "nanoseconds",
+    "microseconds",
+    "milliseconds",
+    "seconds",
+    "minutes",
+    "hours",
+]
+
+_LiteralAbsolute: TypeAlias = Literal[
+    "iso_8601",
+    "epoch_nanoseconds",
+    "epoch_microseconds",
+    "epoch_milliseconds",
+    "epoch_seconds",
+    "epoch_minutes",
+    "epoch_hours",
+]
+
+_LiteralRelativeDeprecated: TypeAlias = Literal[
+    "relative_nanoseconds",
+    "relative_microseconds",
+    "relative_milliseconds",
+    "relative_seconds",
+    "relative_minutes",
+    "relative_hours",
+]
+
+TypedTimeDomain: TypeAlias = Iso8601 | Epoch | Relative | Custom
+_AbsoluteTimeDomain: TypeAlias = Iso8601 | Epoch | Custom | _LiteralAbsolute
+_AnyTimeDomain: TypeAlias = TypedTimeDomain | _LiteralAbsolute | _LiteralRelativeDeprecated
```

Comment:
Python 3.9 doesn't like the mixture of types and literals with a `|` - need to explicitly use `typing.Union`

see https://github.com/python/mypy/issues/11665 - this was already an issue previously in the client lib, just in a different spot.